### PR TITLE
[Merged by Bors] - Make `Transform` propagation correct in the presence of updated children

### LINKED ISSUE #3329: `GlobalTransform` not being updated
## Bevy version

ffecb05a0ab423b46ab991b49c54516bcaf2ea7b

## Operating system & version

Linux

## What you did

Link to repo [here](https://github.com/yilinwei/bevy-issue).

The only file is `src/main.rs`. The bevy submodule is simply to add some extra logging.

## What you expected to happen

The `TransformPlugin` has two systems `parent_update` and `transform_propagate`. `parent_update` is always meant to run before `transform_propagate` [according to the run criteria](https://github.com/bevyengine/bevy/blob/ffecb05a0ab423b46ab991b49c54516bcaf2ea7b/crates/bevy_transform/src/lib.rs#L37).

If an `Entity` has a `Parent` but no `Children`, `parent_update` [creates one](https://github.com/bevyengine/bevy/blob/ffecb05a0ab423b46ab991b49c54516bcaf2ea7b/crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs#L58). The `transform_propagate` then queries from the [root nodes](https://github.com/bevyengine/bevy/blob/ffecb05a0ab423b46ab991b49c54516bcaf2ea7b/crates/bevy_transform/src/transform_propagate_system.rs#L12) and updates the `GlobalTransform`.

Whether you insert a `Children` component [shouldn't matter](https://github.com/yilinwei/bevy-issue/blob/79e48b8a47e8baf0d7f8aac73df3ee6b66ec5f00/src/main.rs#L15). The `Children` should be inserted by the `parent_update` and the `transform_propagate` should pick up the inserted `Children`.

## What actually happened

Whether you insert the `Children` does matter, if you comment the line out you get the `unexpected matrix:` log line, whereas otherwise you don't.

Adding some more `println!` statements, the root node query doesn't pick up `Children` at all - it's empty, but you can clearly see the new `Children` in the later query and adding the `println!` directly where the `Children` is added shows that it does hit that part of the code.

### PR DESCRIPTION
**Pull request by DJMcNab**
Supercedes https://github.com/bevyengine/bevy/pull/3340, and absorbs the test from there.

# Objective

- Fixes #3329

## Solution

- If the `Children` component has changed, we currently do not have a way to know how it has changed.
- Therefore, we must update the hierarchy downwards  from that point to be correct.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
I like this approach much better, thanks.

File: crates/bevy_transform/src/systems.rs [hunk 1]
```diff
@@ -1,10 +1,5 @@
 use crate::components::{GlobalTransform, Transform};
-use bevy_ecs::{
-    entity::Entity,
-    prelude::Changed,
-    query::{With, Without},
-    system::Query,
-};
+use bevy_ecs::prelude::*;
```
Comment (mockersf) [inline crates/bevy_transform/src/systems.rs:2]: Why this change?
The style guide mention preferring granular imports
https://github.com/bevyengine/bevy/blob/7d21ca37446b04ae68f7ee092f545fb6dc2a40fb/.github/contributing/engine_style_guide.md?plain=1#L9

Comment (DJMcNab) [reply crates/bevy_transform/src/systems.rs:2]: I wasn't aware of that guideline. I made this change here because I think I needed something else in the import, and importing the prelude is significantly less prone to merge conflicts overall.

As a note, that style guide seems to only care about cases where you have a dependency on the root `bevy` facade, not when depending directly on the sub crates.

I'm not sure that I really like that guideline - merge conflicts in the imports create a ton of busywork, since the tooling requires a lot of manual handholding.

I can change it back though.

Comment (DJMcNab) [reply crates/bevy_transform/src/systems.rs:2]: I don't like that rule for `bevy_ecs::prelude` - granular imports from other crates are more reasonable (`bevy_app` is marginal)

`bevy_ecs`'s types are the standard vocabulary of bevy, and there should be very few source files which don't use any of them (and where the set used is static). That is, I think glob importing from `bevy_ecs::prelude` should be encouraged. 

I think similarly for `bevy_app::prelude` could be argued, since `bevy_app` and `bevy_ecs` should be merged anyway, although the set used from `bevy_app` is more likely to be static.

**Review by mockersf — APPROVED** [review]

File: crates/bevy_transform/src/systems.rs (hunk 1 above)
Comment (mockersf) [reply crates/bevy_transform/src/systems.rs:2]: > As a note, that style guide seems to only care about cases where you have a dependency on the root bevy facade, not when depending directly on the sub crates.

Yeah, the example in the style guide is wrong, you never import bevy when working in bevy

> I don't like that rule for bevy_ecs::prelude - granular imports from other crates are more reasonable (bevy_app is marginal)
Maybe, but I don't think this PR is the place to change that and I would rather have consistency until we decide to change the rule

**Comment by mockersf** [comment]
Code change is good, I would prefer if the code style would be kept by this PR

File: crates/bevy_transform/src/systems.rs [hunk 2]
```diff
@@ -13,6 +8,7 @@ pub fn transform_propagate_system(
     mut root_query: Query<
         (
             Option<&Children>,
+            Option<Changed<Children>>,
```
Comment (TheRawMeatball) [inline crates/bevy_transform/src/systems.rs:11]: Instead of `Option<Changed<Children>>`, `ChangeTrackers<Children>` should be used.

Comment (DJMcNab) [reply crates/bevy_transform/src/systems.rs:11]: Why? I thought we'd want to support roots without children - this would be quite a serious regression.

I could change it to `Option<ChangeTrackers<Children>>` of course if you prefer, although I would have thought that getting the `changed` only would be easier.

Comment (TheRawMeatball) [reply crates/bevy_transform/src/systems.rs]: Hm, then i suppose using `Option<ChangeTrackers<Children>>` would be better. I'd rather not mix filters in with stuff meant to be queried for, given the effect mixing other filters like `With` can have.

Comment (DJMcNab) [reply crates/bevy_transform/src/systems.rs]: I'm not sure I understand the concern

`Changed` is merely a query which returns true if the component value has changed since the last time the system was run.

Comment (TheRawMeatball) [reply crates/bevy_transform/src/systems.rs]: Maybe, but it's introduced and used as a filter most of the time. I feel like it's bad practice to mix filters with regular queries, given filters like `With` do very different things.

Comment (alice-i-cecile) [reply crates/bevy_transform/src/systems.rs]: I would also like to forbid the use of filters as ordinary query parameters down the line if we can; and this will be one more case that we'd have to revert.

That said, I don't feel strongly, and won't block on it.

File: crates/bevy_transform/src/systems.rs (hunk 1 above)
Comment (alice-i-cecile) [reply crates/bevy_transform/src/systems.rs]: #4644 discusses this change.

**Comment by cart** [comment]
This seems like the right call, but I just created an issue to track fixing this, as the new behavior is not justifiable long term.
